test/zhr_devs/submissions/commands/solution_submitted_test.exs

```diff
@@ -0,0 +1,128 @@
+
+  import Commanded.Assertions.EventAssertions
+
+  import ZhrDevs.Fixtures
+
+  alias Commanded.Aggregates.Aggregate
+
+  alias ZhrDevs.App
+
+  alias ZhrDevs.Submissions.{Aggregates, Commands, Events}
+
+  alias ZhrDevs.Submissions.SubmissionIdentity
+
+  describe "SolutionSubmitted command" do
+    setup do
+      valid_opts = fn identity ->
+        [
+          hashed_identity: DomaOAuth.hash(identity),
+          task_uuid: "Jaju6yAv1oZ23NcjJk-1JkOrxCemsH_K-9iRRw0sYRg=",
+          technology: "elixir",
+          solution_path: "test/support/testfile.txt"
+        ]
+      end
+
+      %{valid_opts: valid_opts}
+    end
+
+    test "return error if invalid options passed" do
+      assert {:error, %KeyError{key: :hashed_identity}} = Commands.SubmitSolution.dispatch([])
+    end
+
+    test "return error if invalid hashed_identity passed" do
+      assert {:error, %ArgumentError{message: "incorrect padding"}} =
+               Commands.SubmitSolution.dispatch(hashed_identity: "invalid")
+    end
+
+    test "return an error when file can't be located (File.exists? is returning false)", %{
+      valid_opts: valid_opts
+    } do
+      opts =
+        identity_generator() |> valid_opts.() |> Keyword.put(:solution_path, "/nonesense/path")
+
+      assert {:error, %ArgumentError{message: "Solution path is invalid: /nonesense/path"}} =
+               Commands.SubmitSolution.dispatch(opts)
+    end
+
+    test "return an error when technology we do not support passed", %{valid_opts: valid_opts} do
+      opts = identity_generator() |> valid_opts.() |> Keyword.put(:technology, "javascript")
+
+      assert {:error, %Uptight.AssertionError{message: message}} =
+               Commands.SubmitSolution.dispatch(opts)
+
+      assert message =~ "javascript is not supported"
+    end
+
+    test "with valid arguments emits an a valid event", %{valid_opts: valid_opts} do
+      identity = identity_generator()
+      opts = valid_opts.(identity)
+
+      assert :ok = Commands.SubmitSolution.dispatch(opts)
+      hashed_identity = opts[:hashed_identity]
+
+      assert_receive_event(
+        App,
+        Events.SolutionSubmitted,
+        fn event -> event.hashed_identity.encoded == hashed_identity end,
+        fn event ->
+          assert %{
+                   solution_path: %Uptight.Text{text: "test/support/testfile.txt"},
+                   task_uuid: %Uptight.Base.Urlsafe{},
+                   uuid: %Uptight.Base.Urlsafe{},
+                   technology: :elixir
+                 } = event
+        end
+      )
+    end
+
+    test "with valid arguments aggregate state is predictable", %{valid_opts: valid_opts} do
+      identity = identity_generator()
+      opts = valid_opts.(identity)
+      hashed_identity = opts[:hashed_identity]
+
+      assert :ok = Commands.SubmitSolution.dispatch(opts)
+
+      wait_for_event(App, Events.SolutionSubmitted, fn event ->
+        event.hashed_identity.encoded == hashed_identity
```

Comment:
Am I right that this is a predicate function and the function will return with timeout or when it receives such an event that it matches the predicate?

If so, your test may pass when this wait function timeouts.

Reply:
This function only returns if the specified event is observed. Otherwise it will raise with timeout.

Reply:
Ok!